web/cobrands/fixmystreet/assets.js

```diff
@@ -16,139 +16,91 @@ var fixmystreet = fixmystreet || {};
     };
 })();
 
-/* Special USRN handling */
-
-(function(){
-
-var selected_usrn = null;
-var usrn_field = null;
+// Handles layers such as USRN, TfL roads, and the like
+OpenLayers.Layer.VectorNearest = OpenLayers.Class(OpenLayers.Layer.Vector, {
+    selected_feature: null,
+
+    initialize: function(name, options) {
+        OpenLayers.Layer.Vector.prototype.initialize.apply(this, arguments);
+        $(fixmystreet).on('maps:update_pin', this.checkFeature.bind(this));
+        $(fixmystreet).on('assets:selected', this.checkFeature.bind(this));
+        $(fixmystreet).on('report_new:category_change:extras_received', this.updateUSRNField.bind(this));
+        $(fixmystreet).on('report_new:category_change', this.changeCategory.bind(this));
+    },
 
-fixmystreet.usrn = {
-    select: function(evt, lonlat) {
-        var usrn_providers = fixmystreet.map.getLayersBy('fixmystreet', {
-            test: function(options) {
-                return options && options.usrn;
-            }
-        });
-        if (usrn_providers.length) {
-            var usrn_layer = usrn_providers[0];
-            usrn_field = usrn_layer.fixmystreet.usrn.field;
-            var point = new OpenLayers.Geometry.Point(lonlat.lon, lonlat.lat);
-            var feature = usrn_layer.getFeatureAtPoint(point);
-            if (feature == null) {
-                // The click wasn't directly over a road, try and find one
-                // nearby
-                feature = usrn_layer.getNearestFeature(point, 10);
-            }
-            if (feature !== null) {
-                selected_usrn = feature.attributes[usrn_layer.fixmystreet.usrn.attribute];
+    checkFeature: function(evt, lonlat) {
+        this.getNearest(lonlat);
+        this.updateUSRNField();
```

Comment:
This working relies, I think, on us only ever having one layer for each USRN, which I expect we would, but it might be worth making that explicit somewhere.

Reply:
I'm not sure what you mean, sorry. You could have two layers with `usrn: { attribute: "...", field: "..." }`  and they would both work independently?

Reply:
But if both of those had the same `field` set then then they might do conflicting things when updating the field.